src/library/scala/collection/concurrent/Map.scala

```diff
@@ -103,6 +103,43 @@ trait Map[K, V] extends scala.collection.mutable.Map[K, V] {
       }
   }
 
+  /**
+   * Removes the entry for the specified key if it's currently mapped to the
+   * specified value. Comparison to the specified value is done using reference
+   * equality.
+   *
+   * Not all map implementations can support removal based on reference
+   * equality, and for those implementations, object equality is used instead.
+   *
+   * $atomicop
+   *
+   * @param k   key for which the entry should be removed
+   * @param v   value expected to be associated with the specified key if
+   *            the removal is to take place
+   * @return    `true` if the removal took place, `false` otherwise
+   */
+  // TODO: make part of the API in a future version
+  private[concurrent] def removeRefEq(k: K, v: V): Boolean = remove(k, v)
```

Comment:
Actually, we should have made this `private[collection]`, as `JConcurrentMapWrapper` is not below `concurrent` (https://github.com/scala/bug/issues/12457). I'll send a PR.

Reply:
does it make a difference if we can't override it meaningfully in `JConcurrentMapWrapper` anyway?

Reply:
or to put it a bit differently, what about if someone writes their own `concurrent.Map` implementation?

Reply:
> what about if someone writes their own `concurrent.Map` implementation?

Not much we can do for them, since the new methods can't be public (or protected).

I think `private[collection]` is just more correct, the compiler should reject `JConcurrentMapWrapper` if its mixin forwarders call `private[concurrent]` methods. But it doesn't make any difference currently.

Reply:
if that's the case, then this change is technically binary incompatible, though not in practice given the current compiler.

Reply:
Ah, great point. If we enfoce the compiler check as I suggested, then a trait `T` with a `private[p]` method could only be subclassed within `p`. This is not what we want.

Reply:
this seems to me like a blocking issue for which we need to figure out a solution before 2.13.7. I'm concerned, however, that we may have already added such `private[p]` methods in a previous release as well.

Reply:
Are you worried about a `private[p]` method leaking out to classes outside `p` in form of a mixin forwarder?

I guess the two cases are: adding and removing a `private[p]` method without recompiling a subclass.

In the first case, the subclass will not have the mixin forwarder. That should be OK, the fowarder is an override of the default method in the interface classfile. Adding / removing overrides is binary compatible. Mixin forwarders are needed to call the correct method (according to Scala linearization) when the same method is inherited from multiple traits. But that shouldn't be the case here.

When removing a `private[p]` trait method, the subclass would remain with a stale mixin forwarder. This should never be invoked though; when the `private[p]` method is removed, all calls to it are removed as well since the package `p` is being re-compiled.

What we ignore, as always, is the possibility to call `private[p]` methods from Java or from external projects that put classes in package `p`.

Reply:
sorry, I misunderstood your initial comment. after re-reading it and reading your comment on the bug ticket, I think I understand, and there's no issue (blocking or otherwise).

--------

the only other concern I have (which is separate from what you mentioned) is the following scenario:

```scala
// A.scala version 1
package p

trait A
```

```scala
// B.scala, compiled against A.scala version 1
package q

class B extends p.A {
  def foo: Int = 1
}
```

```scala
// A.scala version 2
package p

trait A {
  private[p] def foo: Int = 2
}
```

a trait `A` adds a `private[p]` method with the same signature as one already defined in a subclass `B`, compiled against a version of `A` without that method. what happens in that case?

Reply:
Yeah, then the method `B.foo` is an override of `A.foo` at run-time. The same happens if `A` is a `class`. I guess this is a limitation we can live with.